packages/playwright/src/runner/watchMode.ts

```diff
@@ -19,25 +19,77 @@ import { createGuid, getPackageManagerExecCommand, ManualPromise } from 'playwri
-import ListReporter from '../reporters/list';
+import { TestServerDispatcher } from './testServer';
+import { EventEmitter } from 'stream';
+import { type TestServerSocket, TestServerConnection } from '../isomorphic/testServerConnection';
 
-class FSWatcher {
-  private _dirtyTestFiles = new Map<FullProjectInternal, Set<string>>();
-  private _notifyDirtyFiles: (() => void) | undefined;
-  private _watcher: CFSWatcher | undefined;
-  private _timer: NodeJS.Timeout | undefined;
+class InMemoryServerSocket extends EventEmitter implements TestServerSocket {
+  constructor(public readonly send: (data: any) => void, public readonly close: () => void = () => {}) {
+    super();
+  }
+
+  addEventListener(event: string, listener: (e: any) => void) {
+    this.addListener(event, listener);
+  }
+}
+
+export async function runWatchModeLoop(config: FullConfigInternal): Promise<FullResult['status']> {
+  const testServerDispatcher = new TestServerDispatcher({ configDir: config.configDir, resolvedConfigFile: config.config.configFile });
+  const inMemorySocket = new InMemoryServerSocket(
+      async data => {
+        const { id, method, params } = JSON.parse(data);
+        try {
+          const result = await testServerDispatcher.transport.dispatch(method, params);
+          inMemorySocket.emit('message', { data: JSON.stringify({ id, result }) });
+        } catch (e) {
+          inMemorySocket.emit('message', { data: JSON.stringify({ id, error: String(e) }) });
+        }
+      }
+  );
+  testServerDispatcher.transport.sendEvent = (method, params) => {
+    inMemorySocket.emit('message', { data: JSON.stringify({ method, params }) });
+  };
+  const testServerConnection = new TestServerConnection(inMemorySocket);
+  inMemorySocket.emit('open');
+
+  const dirtyTestFiles = new Map<FullProjectInternal, Set<string>>();
+  const onDirtyTestFiles: { resolve?(): void } = {};
+  const failedTestIdCollector = new Set<string>();
+
+  testServerConnection.onTestFilesChanged(({ testFiles }) => onTestFilesChanged(testFiles));
+  testServerConnection.onReport(report => {
+    if (report.method === 'onTestEnd') {
+      const { result: { status }, test: { testId, expectedStatus } } = report.params;
+      if (status !== expectedStatus)
+        failedTestIdCollector.add(testId);
+      else
+        failedTestIdCollector.delete(testId);
+    }
+  });
+
+  const originalStdoutWrite = process.stdout.write;
+  const originalStderrWrite = process.stderr.write;
+  testServerConnection.onStdio(chunk => {
+    const text = chunk.text ?? Buffer.from(chunk.buffer!, 'base64');
+    if (chunk.type === 'stderr')
+      originalStderrWrite.call(process.stderr, text);
+    else
+      originalStdoutWrite.call(process.stdout, text);
+  });
+
+  await testServerConnection.initialize({ interceptStdio: true });
+  await testServerConnection.runGlobalSetup({});
+
+  await testServerConnection.listTests({});
+  await testServerConnection.watch({ fileNames: [] });
```

Comment:
It seems like watch mode and ui mode have different ideas about watch semantics. Let's write them down and figure out whether we want to unify them or maybe update test server to support both, or something else.

Reply:
The big difference I see is that in `--watch`, there's no way of selecting individual parts of the test tree to be watched. I also don't think that'd be ergonomic in the terminal. Are there any other differences that we want to unify?

Reply:
- UI mode watches a specific list of files, and passes all test files when "Watch All" mode is enabled, while this watch mode does not "watch" any files.
- UI mode does not run dependencies upon a watched test run, while this mode does.
- This mode applies filters like grep after being notified about changed files, while UI mode applies filters before passing the list of files to watch (IIUC).

Reply:
- In the current implementation, I don't see the UI calling `watch` at all. It calls `initialize` once with `watchTestDirs: true`, but apart from that it never changes the list of watched files. I made this implementation do the same in c6d72fe2a1e061acdfee64a59eea9c61d3b4208b.

- UI mode does run dependencies, see 184b2b03eb4c7e02d52824546efa39b2cdbf6784 for a test that proves that.

- UI mode applies the filters after being notified as well. It does that in https://github.com/Skn0tt/playwright/blob/184b2b03eb4c7e02d52824546efa39b2cdbf6784/packages/trace-viewer/src/ui/uiModeView.tsx#L352-L374.